### PR TITLE
Blog post about 'Science on Rails'

### PR DESCRIPTION
**Pull request by gvwilson**
Comments welcomed.

### CODE REVIEW

File: blog/2013/06/twelve-bar-blues-of-science.html [hunk 1]
```diff
@@ -0,0 +1,134 @@
...
+    I mentioned <code>rails new project</code> a moment ago;
+    Rails, Django, and other frameworks of the same ilk
+    create a command application for each project
+    so that people can add new features in a reliable, findable way with a few keystrokes.
+    Science projects should have something similar for adding new data sets, producing new results, etc.
+  </li>
+  <li>
+    <em>Pluginnable.</em>
+    Rails didn't just put the application code and data in particular places;
+    it also provided slots for adding new management commands and analysis tools.
```
Comment (ptone) [inline blog/2013/06/twelve-bar-blues-of-science.html:109]: talking about "analysis tools" here may be leaking your science into the rails metaphor

**Comment by ptone** [comment]
[this turned into a longer comment than intended - should be posted on the published version instead, but I'll post here FWIW]

This is good stuff - while the convention over configuration pattern was one aspect of the success of Rails, it can't explain the boon in web frameworks, and frameworks in general.

Frameworks handle the parts that are repetitive and extremely common, and also those parts that can be hard to get right (ie password storage).

Web frameworks were able to operate in an environment that was, for the most part, already standardized around HTTP.

Another observation about software frameworks vs software is that things like Rails or Django don't do anything by themselves out of the box, they are authoring frameworks, the product is a custom web app, while software can be used simply as a tool with no customization.

Part of the challenge of coming up with the story for new convention empowered science is deciding what the "product" becomes when it is no longer just "papers".

Something analogous to "releases" will likely be have to be put in place. Even the most rabidly evangelical open source programmers will usually develop new projects initially "in private", sharing initially with some closed friends for input, but as early as possible publishing dev, alpha, beta, RC "releases".

The scaffolding of new projects is one of the other things you mentioned, is one of the more useful "convention" tools. And Django somewhat recently introduced the idea of project templates that the community could release.  Speaking of community - key to the success of these developer communities has been the trifecta of Community (IRC, conventions, Mailing lists etc), Documentation, Open Source.

Hopefully there will be a way to build the solution on the Unix model, with small independent tools - as trying to write an Xcode for science will fail. Another warning tale is this:
http://hueniverse.com/2012/07/oauth-2-0-and-the-road-to-hell/

In this case, the endless debate among big institutions can kill efforts - I'm more intrigued by the early-days Facebook model, build something useful, let it spread like wildfire if people like it.

File: blog/2013/06/twelve-bar-blues-of-science.html [hunk 2]
```diff
@@ -0,0 +1,134 @@
...
+  Many kinds of scientific work are equally well structured:
+  the results aren't predictable&mdash;it wouldn't be research if they were&mdash;but
+  the equipment setup, sample preparation, note-taking, statistics, and write-up
+  are (mostly) structured in ways that other scientists are familiar with.
+  This lets them pick up each other's projects more quickly;
+  it also gives scientists more time to do what's unique about a particular project
+  because they don't have to spend time thinking about the things that aren't.
+</p>
+<p>
+  Open science (or digital research, or whatever you want to call it)
```
Comment (jiffyclub) [inline blog/2013/06/twelve-bar-blues-of-science.html]: Does "open science" necessarily mean "shared via the web"? Want to use a more specific term here?

File: blog/2013/06/twelve-bar-blues-of-science.html [hunk 3]
```diff
@@ -0,0 +1,134 @@
...
+  <li>
+    <em>Standard commands to tie it all together.</em>
+    I mentioned <code>rails new project</code> a moment ago;
+    Rails, Django, and other frameworks of the same ilk
+    create a command application for each project
+    so that people can add new features in a reliable, findable way with a few keystrokes.
+    Science projects should have something similar for adding new data sets, producing new results, etc.
+  </li>
+  <li>
+    <em>Pluginnable.</em>
```
Comment (jiffyclub) [inline blog/2013/06/twelve-bar-blues-of-science.html]: Sub "plugin compatible"?

File: blog/2013/06/twelve-bar-blues-of-science.html [hunk 4]
```diff
@@ -0,0 +1,134 @@
...
+    Rails didn't just put the application code and data in particular places;
+    it also provided slots for adding new management commands and analysis tools.
+  </li>
+  <li>
+    <em>Forkable and mergeable.</em>
+    Whatever Rails-like system scientists use <em>must</em> place nicely with version control,
```
Comment (jiffyclub) [inline blog/2013/06/twelve-bar-blues-of-science.html]: s/place/play/

File: blog/2013/06/twelve-bar-blues-of-science.html [hunk 5]
```diff
@@ -0,0 +1,134 @@
...
+  </li>
+  <li>
+    <em>Federated.</em>
+    The web lets us access information stored anywhere on the planet;
+    ironically, our response so far has been to centralize like never before.
+    Wikipedia, Facebook, and GitHub are effectively all single points of failure:
+    one court order,
+    change of policy,
+    or (as we recently discovered with Mendeley) ideologically incompatible acquisition
+    can have wide-reaching impact.
```
Comment (jiffyclub) [inline blog/2013/06/twelve-bar-blues-of-science.html:129]: can have _a_ wide-reaching impact

File: blog/2013/06/twelve-bar-blues-of-science.html (hunk 3 above)
Comment (ptone) [inline blog/2013/06/twelve-bar-blues-of-science.html]: I think "pluggable" or "pluggable design" is the most common term for this concept

File: blog/2013/06/twelve-bar-blues-of-science.html [hunk 6]
```diff
@@ -0,0 +1,134 @@
...
+    they will fork and merge projects.
+  </li>
+  <li>
+    <em>Federated.</em>
+    The web lets us access information stored anywhere on the planet;
+    ironically, our response so far has been to centralize like never before.
+    Wikipedia, Facebook, and GitHub are effectively all single points of failure:
+    one court order,
+    change of policy,
+    or (as we recently discovered with Mendeley) ideologically incompatible acquisition
```
Comment (gdevenyi) [inline blog/2013/06/twelve-bar-blues-of-science.html]: A link re: Mendeley perhaps?

http://www.newyorker.com/online/blogs/elements/2013/04/elsevier-mendeley-journals-science-software.html

**Comment by timClicks** [comment]
I don't know whether this is appropriate here, or if it would be better placed as a comment to the actual blog post when it appears.

### What excited me about Ruby on Rails several years ago?

There are some other things which I think really impressed me with Rails when I first encountered it, which I don't know have been covered by your post:
- The "create a blog in 15 minutes" webcast was mind-blowing. Here was someone starting from nothing and getting a functional database-driven, dynamic website up and running.  
- Rails was pretty. Even its error pages were attractive.

### What lessons could Science on Rails learn from this?

New users need to be able to get something tangible from the system quickly. 

Invest time in design. In order for a project this ambitious to be successful, it needs to be marketable. 

Would Science on Rails be able to cater for the needs of three forms of knowledge discovery, e.g. theoretical, experimental and also simulation/modelling?

### Other places to learn from

One other project that we could learn from is [Quickly](https://wiki.ubuntu.com/Quickly) from the Ubuntu community. They want to make it easier for people to create applications (that run on Ubuntu) by providing a rich set of templates, a standard database and so forth.

I actually wonder if Science on Rails could start its life as a Quickly template. 

### Tools

It's completely the wrong time for this, but I can't help but add some thoughts about what tools would be suitable given the goals of being able to make it easy to fork and merge.

I think CouchDB and/or SQLite would be the best database(s) to work from. This is due to the simplicity of replication (e.g. fork & merge) at the data level. For CouchDB, syncing is a matter of entering a URL into a web console. For SQLite, it's running a single file.

For small teams this might even be running SQLite in a Dropbox shared folder.

**Comment by khinsen** [comment]
It's important not to forget the dangers of applying Web attitudes to science. In particular, the Web is known for fast and chaotic changes (both technological and organiztional), whereas science progresses slowly, always building on results from the past, including the distant past. I regularly cite papers that are 50 years old.

You mentioned one aspect along these lines, the risk of centralization creating single points of failure. But in a federated system, projects must be able to survive the disappearance of their hosting site. It must therefore be possible to transfer a project from one host to another, without losing anything in the process. That requires standards for data exchange and storage. Archiving finished projects has much the same requirements, with careful documentation of all data formats thrown in for the benefit of future scientists. Anyone setting up a "Web for science" should talk to a documentalist about such issues. Shameless self-advertising: http://dx.doi.org/10.1109/MCSE.2012.108 (no open acces, sorry)

**Pull request merged**